File: pbwebwcl/src/app/checklist/services/checkList.service.ts
/*
 * Copyright Regione Piemonte - 2024 
 * SPDX-License-Identifier: EUPL-1.2 
*/

import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { ConfigService } from 'src/app/core/services/config.service';
import { Observable } from 'rxjs';
import { ChecklistHtmlDTO } from 'src/app/shared/commons/dto/checklisthtml-dto';
import { CodiceDescrizioneDTO } from '@pbandi/common-lib';
import { CheckListItem } from '../commons/dto/CheckListItem';
import { InizializzaGestioneChecklistDTO } from '../commons/dto/inizializza-gestione-checklist-dto';
import { EsitoDTO } from 'src/app/shared/commons/dto/esito-dto';
import { EsitoLockCheckListDTO } from '../commons/dto/esito-lock-checklist-dto';
import { EsitoOperazioneDTO } from '../commons/dto/esito-operazione-dto';
import { UserInfoSec } from 'src/app/core/commons/dto/user-info';
import { CercaChecklistRequestDTO } from '../commons/dto/cerca-checklist-request-dto';

@Injectable()
export class CheckListService {


    constructor(
        private configService: ConfigService,
        private http: HttpClient
    ) { }


    getModuloCheckList(idProgetto: number): Observable<ChecklistHtmlDTO> {
        let url = this.configService.getApiURL() + "/restfacade/checklist/getModuloCheckList";
        let params = new HttpParams().set("idProgetto", idProgetto.toString());

        console.log('ANG CheckListService getModuloCheckList ,idProgetto =' + idProgetto);

        return this.http.get<ChecklistHtmlDTO>(url, { params: params });
    }

    getCheckListInLocoHtml(idDocumentoIndex: number): Observable<ChecklistHtmlDTO> {
        let url = this.configService.getApiURL() + "/restfacade/checklist/getCheckListInLocoHtml";
        let params = new HttpParams().set("idDocumentoIndex", idDocumentoIndex.toString());
        return this.http.get<ChecklistHtmlDTO>(url, { params: params });
    }

    getCheckListValidazioneHtml(idDocumentoIndex: number, idDichiarazione: number, idProgetto: number) {
        let url = this.configService.getApiURL() + "/restfacade/checklist/getCheckListValidazioneHtml";
        let params = new HttpParams().set("idDocumentoIndex", idDocumentoIndex.toString())
            .set("idDichiarazione", idDichiarazione.toString())  //idTarget
            .set("idProgetto", idProgetto.toString());
        return this.http.get<ChecklistHtmlDTO>(url, { params: params });
    }

    /*
     getModelloCheckListAffidamentoHtml(idProgetto: number, idAffidamento: number, soggettoControllore: string, codTipoChecklist: string): Observable<ChecklistHtmlDTO> {
            let url = this.configService.getApiURL() + "/restfacade/checklist/modelloAffidamentoHtml";
            let params = new HttpParams().set("idProgetto", idProgetto.toString())
                                         .set("idAffidamento", idAffidamento.toString())
                                         .set("soggettoControllore", soggettoControllore.toString())
                                         .set("codTipoChecklist", codTipoChecklist.toString());
    
            return this.http.get<ChecklistHtmlDTO>(url, { params: params });
        }
    */

    caricaDichiarazioneDiSpesa(idProgetto: number, isFP: boolean): Observable<Array<CodiceDescrizioneDTO>> {
        let url = this.configService.getApiURL() + "/restfacade/checklist/caricaDichiarazioneDiSpesa";
        let params = new HttpParams().set("idProgetto", idProgetto.toString()).set("isFP", isFP ? "S" : "N");

        console.log('ANG CheckListService caricaDichiarazioneDiSpesa ,idProgetto =' + idProgetto);
        return this.http.get<Array<CodiceDescrizioneDTO>>(url, { params: params });
    }

    caricaStatoSoggetto(idUtente: number, idIride: string): Observable<Array<CodiceDescrizioneDTO>> {
        let url = this.configService.getApiURL() + "/restfacade/checklist/caricaStatoSoggetto";
        let params = new HttpParams().set("idUtente", idUtente.toString())
            .set("idIride", idIride.toString());

        console.log('ANG CheckListService caricaStatoSoggetto ,idUtente =' + idUtente + ', idIride=' + idIride);
        return this.http.get<Array<CodiceDescrizioneDTO>>(url, { params: params });
    }

    cercaChecklist(dichiarazioneSpesa: number, dataControllo: string, soggettoControllo: string,
        stati: string[], tipologia: string, rilevazione: string, versione: string, idProgetto: number): Observable<Array<CheckListItem>> {

        let url = this.configService.getApiURL() + "/restfacade/checklist/cercaChecklist";
        let request = new CercaChecklistRequestDTO(dichiarazioneSpesa?.toString() || null, dataControllo, soggettoControllo, stati, tipologia, rilevazione, versione,
            idProgetto.toString() || null);

        console.log('ANG CheckListService cercaChecklist ,dichiarazioneSpesa =' + dichiarazioneSpesa);
        return this.http.post<Array<CheckListItem>>(url, request);
    }

    inizializzaGestioneChecklist(idProgetto: number): Observable<InizializzaGestioneChecklistDTO> {
        let url = this.configService.getApiURL() + "/restfacade/checklist/inizializzaGestioneChecklist";
        let params = new HttpParams().set("idProgetto", idProgetto.toString());
        return this.http.get<InizializzaGestioneChecklistDTO>(url, { params: params });
    }

    eliminaChecklist(idDocumentoIndex: number): Observable<EsitoDTO> {
        let url = this.configService.getApiURL() + "/restfacade/checklist/eliminaChecklist";
        let params = new HttpParams().set("idDocumentoIndex", idDocumentoIndex.toString());
        return this.http.get<EsitoDTO>(url, { params: params });
    }

    salvaLockCheckListInLoco(idProgetto: number, idDocumentoIndex: number): Observable<EsitoLockCheckListDTO> {
        let url = this.configService.getApiURL() + "/restfacade/checklist/salvaLockCheckListInLoco";
        let params = new HttpParams().set("idProgetto", idProgetto.toString()).set("idDocumentoIndex", idDocumentoIndex.toString());
        return this.http.get<EsitoLockCheckListDTO>(url, { params: params });
    }

    salvaLockCheckListValidazione(idProgetto: number, idDichiarazione: number): Observable<EsitoLockCheckListDTO> {
        let url = this.configService.getApiURL() + "/restfacade/checklist/salvaLockCheckListValidazione";
        let params = new HttpParams().set("idProgetto", idProgetto.toString()).set("idDichiarazione", idDichiarazione.toString());
        return this.http.get<EsitoLockCheckListDTO>(url, { params: params });
    }

    saveCheckListInLocoHtml(statoCheckList: string, checklistHtml: string, idChecklist: string, idProgetto: number, userInfo: UserInfoSec): Observable<EsitoOperazioneDTO> {
        let url = this.configService.getApiURL() + "/restfacade/checklist/salvaCheckListInLoco";
        let formData = new FormData();
        formData.append("idUtente", userInfo.idUtente.toString());
        formData.append("idIride", userInfo.idIride);

        formData.append("statoChecklist", statoCheckList);
        formData.append("checklistHtml", checklistHtml);
        formData.append("idChecklist", idChecklist);
        formData.append("idProgetto", idProgetto.toString());
        /*
                let params = new HttpParams().set("statoChecklist", statoChecklist)
                                             .set("checklistHtml", checklistHtml)
                                             .set("idChecklist", idChecklist)
                                             .set("idProgetto", idProgetto.toString());
        */
        return this.http.post<EsitoOperazioneDTO>(url, formData); //{ params: params });
    }

    saveCheckListDocumentaleHtml(statoCheckList: string, checklistHtml: string, idChecklist: number, idProgetto: number,
        userInfo: UserInfoSec, files: Array<File>): Observable<EsitoOperazioneDTO> {
        let url = this.configService.getApiURL() + "/restfacade/checklist/saveCheckListDocumentaleHtml";
        let formData = new FormData();
        formData.append("idUtente", userInfo.idUtente.toString());
        formData.append("idIride", userInfo.idIride);

        formData.append("statoChecklist", statoCheckList);
        formData.append("checklistHtml", checklistHtml);
        formData.append("idChecklist", idChecklist.toString());
        formData.append("idProgetto", idProgetto.toString());
        if (files) {
            for (var i = 0; i < files.length; i++) {
                formData.append("file", files[i], files[i].name);
            }
        }

        return this.http.post<EsitoOperazioneDTO>(url, formData);
    }

    saveCheckListDefinitivo(statoCheckList: string, checklistHtml: string, idChecklist: number, idProgetto: number,
        userInfo: UserInfoSec, files: File[]): Observable<EsitoOperazioneDTO> {
        let url = this.configService.getApiURL() + "/restfacade/checklist/saveCheckListDefinitivo";
        let formData = new FormData();
        formData.append("idUtente", userInfo.idUtente.toString());
        formData.append("idIride", userInfo.idIride);

        formData.append("statoChecklist", statoCheckList);
        formData.append("checklistHtml", checklistHtml);
        if (idChecklist == null) {
            //formData.append("idChecklist", null);
        } else {
            formData.append("idChecklist", idChecklist.toString());
        }
        formData.append("idProgetto", idProgetto.toString());

        for (let i = 0; i < files.length; i++) {
            formData.append("file", files[i], files[i].name);
        }
        return this.http.post<EsitoOperazioneDTO>(url, formData);
    }

    //PBSERVIZIT
    allegaFilesChecklist(codTipoDocumento: string, idDocumentoIndexChecklist: number, idProgetto: number,
        files: File[]): Observable<EsitoOperazioneDTO> {
        let url = this.configService.getApiURL() + "/restfacade/checklist-common/allegaFilesChecklist";
        let formData = new FormData();
        formData.append("codTipoDocumento", codTipoDocumento);
        formData.append("idDocumentoIndexChecklist", idDocumentoIndexChecklist.toString());
        formData.append("idProgetto", idProgetto.toString());
        for (let i = 0; i < files.length; i++) {
            formData.append("file", files[i], files[i].name);
        }
        return this.http.post<EsitoOperazioneDTO>(url, formData);
    }


}

File: pbweb/pbweb-jar/src/main/java/it/csi/pbandi/pbweb/business/service/CheckListService.java
/*******************************************************************************
* Copyright Regione Piemonte - 2024
* SPDX-License-Identifier: EUPL-1.2
******************************************************************************/

package it.csi.pbandi.pbweb.business.service;

import java.security.InvalidParameterException;

import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.Consumes;
import javax.ws.rs.DefaultValue;
import javax.ws.rs.FormParam;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.QueryParam;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import org.jboss.resteasy.plugins.providers.multipart.MultipartFormDataInput;

import it.csi.pbandi.pbweb.dto.CercaChecklistRequestDTO;



@Path("/checklist")
public interface CheckListService {
		
	@GET
	@Path("getModuloCheckList")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getModuloCheckList( @QueryParam("idProgetto") String idProgetto, 
			@Context HttpServletRequest req) 
					throws InvalidParameterException, Exception;
	
	@GET
	@Path("getCheckListInLocoHtml")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getCheckListInLocoHtml( @QueryParam("idDocumentoIndex") String idDocumentoIndex, 
			@Context HttpServletRequest req) 
					throws InvalidParameterException, Exception;
	
	@GET
	@Path("getCheckListValidazioneHtml")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getCheckListValidazioneHtml( @QueryParam("idDocumentoIndex") String idDocumentoIndex, 
			@QueryParam("idDichiarazione") String idDichiarazione,
			@QueryParam("idProgetto") String idProgetto,
			@Context HttpServletRequest req) 
					throws InvalidParameterException, Exception;

	
	/**
	 * @GET
	@Path("tipologieDocumento") 
	@Produces(MediaType.APPLICATION_JSON)
	Response ottieniTipologieDocumentiDiSpesa(@DefaultValue("0") @QueryParam("idBandoLinea") int idBandoLinea);
	 */
	
	
	@GET
	@Path("caricaDichiarazioneDiSpesa")
	@Produces(MediaType.APPLICATION_JSON)
	public Response caricaDichiarazioneDiSpesa(
			@QueryParam("idProgetto") String idProgetto,
			@QueryParam("isFP") String isFP,
			@Context HttpServletRequest req) throws InvalidParameterException, Exception;
	
	@GET
	@Path("caricaStatoSoggetto")
	@Produces(MediaType.APPLICATION_JSON)
	public Response caricaStatoSoggetto( @QueryParam("idUtente") String idUtente, 
										@QueryParam("idIride") String idIride, 
										@Context HttpServletRequest req) 
					throws InvalidParameterException, Exception;
	
	
	@POST
	@Path("cercaChecklist")
	@Produces(MediaType.APPLICATION_JSON)
	public Response cercaChecklist( CercaChecklistRequestDTO request,
									@Context HttpServletRequest req) 
					throws InvalidParameterException, Exception;
	
	@GET
	@Path("inizializzaGestioneChecklist")
	@Produces(MediaType.APPLICATION_JSON)
	public Response inizializzaGestioneChecklist(@QueryParam("idProgetto") Long idProgetto,
												 @Context HttpServletRequest req) 
					throws InvalidParameterException, Exception;
	
	@GET
	@Path("eliminaChecklist")
	@Produces(MediaType.APPLICATION_JSON)
	public Response eliminaChecklist(@QueryParam("idDocumentoIndex") Long idDocumentoIndex, 
									 @Context HttpServletRequest req) 
					throws InvalidParameterException, Exception;
	
	@GET
	@Path("salvaLockCheckListInLoco")
	@Produces(MediaType.APPLICATION_JSON)
	public Response salvaLockCheckListInLoco(
			@QueryParam("idProgetto") Long idProgetto, 
			@QueryParam("idDocumentoIndex") Long idDocumentoIndex,
			@Context HttpServletRequest req) 
					throws InvalidParameterException, Exception;
	
	@GET
	@Path("salvaLockCheckListValidazione")
	@Produces(MediaType.APPLICATION_JSON)
	public Response salvaLockCheckListValidazione(
			@QueryParam("idProgetto") Long idProgetto, 
			@QueryParam("idDichiarazione") Long idDichiarazione,
			@Context HttpServletRequest req) 
					throws InvalidParameterException, Exception;
	
	@POST
	@Path("salvaCheckListInLoco")
	@Produces(MediaType.APPLICATION_JSON)
	@Consumes(MediaType.MULTIPART_FORM_DATA)
	public Response salvaCheckListInLoco( MultipartFormDataInput multipartFormData) 
					throws InvalidParameterException, Exception;
	
	@POST
	@Path("saveCheckListDocumentaleHtml")
	@Produces(MediaType.APPLICATION_JSON)
	@Consumes(MediaType.MULTIPART_FORM_DATA)
	public Response saveCheckListDocumentaleHtml( MultipartFormDataInput multipartFormData) 
					throws InvalidParameterException, Exception;
	
	
	// Chiamato quando da Gestione Checklist si salva una checklist in loco definitiva (bottone 'salva definitiva').
	@POST
	@Path("saveCheckListDefinitivo")
	@Produces(MediaType.APPLICATION_JSON)
	@Consumes(MediaType.MULTIPART_FORM_DATA)
	public Response saveCheckListDefinitivo(  MultipartFormDataInput multipartFormData) 
					throws InvalidParameterException, Exception;

}
